### PR TITLE
storage: Support for VDO

### PR DESCRIPTION
**Pull request by mvollmer**
- [x] Default the logical size to physical size during creation
- [x] Allow setting index memory size during creation
- [x] Compression / dedup checkboxes during creation
- [x] Show index size on disk and in memory
- [x] Show space used in GiB
- [x] Don't support partitions
- [x] Allow growing of logical and physical sizes
- [x] Grow filesystem with volume
- [x] #7955 
- [x] #8004
- [x] Tests 
- [x] Feature check should be more quiet when vdo isn't there.
- [x] Errors during delete seem to be swallowed
- [x] Need to wipe block device before creation
- [x] Detect volumes in intermediate states and offer to forcefully remove them
- [x] Link from content tab to vdo doesn't work
- [x] #8062 
- [x] Video 

Video: https://www.youtube.com/watch?v=_iOYN4Y24aY

### CODE REVIEW

**Comment by mvollmer** [comment]
There are some generally useful goodies here that I should probably factor out into their own pull requests.

**Comment by petervo** [comment]
> There are some generally useful goodies here that I should probably factor out into their own pull requests.

Yeah lets get these in right away

File: pkg/storaged/vdo-details.jsx [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+                              Title: _("Delete"),
+                              Danger: _("Deleting a VDO device will erase all data on it."),
+                              action: function () {
+                                  return teardown_active_usage(client, usage)
+                                      .then(function () {
+                                          return teardown_configs()
+                                              .then(function () {
+                                                  var location = cockpit.location;
+                                                  return vdo.remove().then(function () {
+                                                      location.go("/");
```
Comment (petervo) [inline pkg/storaged/vdo-details.jsx]: Don't we want this to go the storage home page rather than /?

File: test/verify/check-storage-vdo [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+from storagelib import *
+
+class TestStorage(StorageCase):
+    def testVdo(self):
+        m = self.machine
+        b = self.browser
+
+        self.login_and_go("/storage")
+
+        if m.execute("which vdo || true") == "":
```
Comment (petervo) [inline test/verify/check-storage-vdo]: Should this be a decorator?

File: pkg/storaged/client.js [hunk 3]
```diff
@@ -525,6 +540,171 @@
...
+            self.volumes = data.map(function (vol, index) {
+                var name = vol.name;
+
+                function volcmd(args) {
+                    return cmd([ "--name", name ].concat(args));
+                }
+
+                var v = { name: name,
+                          broken: vol.broken,
+                          dev: "/dev/mapper/" + name,
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/client.js]: This is always going to be `/dev/mapper`, unless people install new udev rules. (But if they do that, we don't have any way to know what the new prefix actually is.)

File: pkg/storaged/vdo-details.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/storaged/vdo-details.jsx]: `cockpit.location.go` does local navigation and "/" is actually the storage home.  To navigate to the global "/", one needs to use `cockpit.jump`, I think.

File: test/verify/check-storage-vdo (hunk 2 above)
Comment (mvollmer) [reply test/verify/check-storage-vdo]: Hmm, the test actually checks that the VDO panel does not appear on images that don't have the vdo utilities.  We do this for other optional features as well, such as iSCSI.

Comment (petervo) [reply test/verify/check-storage-vdo]: Right, i didn't mean to skip. I just saw it was repeated everywhere, but probably not a big deal.

File: pkg/storaged/vdo-details.jsx (hunk 1 above)
Comment (petervo) [reply pkg/storaged/vdo-details.jsx]: You are right.

File: pkg/storaged/client.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (petervo) [reply pkg/storaged/client.js]: OK makes sense.

**Comment by martinpitt** [comment]
I did a (relatively coarse) review, and I didn't see major issues. This [failure on rhel-7-5](http://209.132.184.41/logs/pull-7817-20171212-084646-7f98ba5a-verify-rhel-7-5/log.html#127) seems relevant:

    vdo: ERROR - dd: failed to open \u2018/dev/sda\u2019: Invalid argument

The other failures are the usual suspect flakes, retrying one round.

**Comment by martinpitt** [comment]
Retried it again, still the same failure on rhel-7-5

**Comment by mvollmer** [comment]
vdo: ERROR - dd: failed to open \u2018/dev/sda\u2019: Invalid argument

The test case used a hard coded vdo configuration for /dev/sda, but didn't have an actual /dev/sda.  That is probably the problem, and should be fixed.

**Comment by mvollmer** [comment]
> and should be fixed.

There is a new failure after I 'corrected' the other test...

**Comment by mvollmer** [comment]
> There is a new failure after I 'corrected' the other test...

Should also be fixed now.

**Pull request merged**